### PR TITLE
apiserver: self-signed in-memory cert for loopback

### PR DESCRIPTION
**Pull request by sttts**
We had complicated logic before to find a matching (SNI- or server-) cert for the loopback device. In the worst case, bootstrapping failed because the provided certs didn't have the binding IP or localhost.

We now pass an artifical `ServerName` "apiserver-loopback-client" with the loopback client and install a self-signed in-memory cert+key into the server for that hostname. The client accepts the corresponding cert.

Fixing https://github.com/kubernetes/kubernetes/issues/41081

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -86,6 +86,8 @@ type Config struct {
 	// Server should be accessed without verifying the TLS
 	// certificate. For testing only.
 	Insecure bool
+	// ServerName is an optional override passed to the server for SNI and used to check server ceritificates against.
+	ServerName string
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go]: having both Host and ServerName is a little crazy... can this go in TLSClientConfig or be named TLSServerName?

Comment (sttts) [inline staging/src/k8s.io/client-go/rest/config.go]: Agree. Will see what makes more sense.

Comment (sttts) [inline staging/src/k8s.io/client-go/rest/config.go]: Fixed.

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -132,6 +128,12 @@ type ImpersonationConfig struct {
 
 // TLSClientConfig contains settings to enable transport layer security
 type TLSClientConfig struct {
+	// Server should be accessed without verifying the TLS certificate. For testing only.
+	Insecure bool
+	// ServerName is an optional override passed to the server for SNI and used to check
+	// server ceritificates against.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go]: certificates

Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go]: if empty, the hostname used to contact the server is used

File: staging/src/k8s.io/client-go/rest/config.go [hunk 3]
```diff
@@ -83,10 +83,6 @@ type Config struct {
 	// TLSClientConfig contains settings to enable transport layer security
 	TLSClientConfig
 
-	// Server should be accessed without verifying the TLS
-	// certificate. For testing only.
-	Insecure bool
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go]: surprised the ripples of moving this weren't worse

Comment (sttts) [inline staging/src/k8s.io/client-go/rest/config.go]: They were. Missed some. Repushed.

File: staging/src/k8s.io/client-go/rest/config.go (hunk 2 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/rest/config.go]: fixed

**Comment by sttts** [comment]
@k8s-bot bazel test this issue #40295

**Comment by sttts** [comment]
@liggitt any more comments?
@smarterclayton approved?

**Comment by liggitt** [comment]
this LGTM, though I'm unsure whether the go-client staging changes are supposed to be made by hand like this

**Comment by sttts** [comment]
All the client-go files are authorative. Only the files in pkg are not.

Best regards,
   Stefan

**Comment by liggitt** [comment]
/lgtm
needs rebase

**Comment by sttts** [comment]
@k8s-bot bazel test this

**Comment by sttts** [comment]
@k8s-bot bazel test this

**Comment by sttts** [comment]
@k8s-bot gci gce e2e test this

**Comment by deads2k** [comment]
This broke the, "generate cert and read generated cert flow" used by this test: https://github.com/kubernetes/kubernetes/pull/42059 .  Haven't sorted out how yet.

**Pull request merged**